bus/pubsub/consumer.go

```diff
@@ -0,0 +1,131 @@
+package pubsub
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"time"
+
+	"cloud.google.com/go/pubsub"
+	"github.com/pcelvng/task/bus/info"
+	"google.golang.org/api/option"
+)
+
+type Consumer struct {
+	// Client is a Google Pub/Sub client scoped to a single project.
+	// Clients should be reused rather than being created as needed.
+	// A Client may be shared by multiple goroutines. Consumer Client
+	client *pubsub.Client
+	sub    *pubsub.Subscription
+
+	// context for clean shutdown
+	ctx  context.Context
+	cncl context.CancelFunc
+
+	info info.Consumer
+}
+
+func (o *Option) NewConsumer() (c *Consumer, err error) {
+	opts := make([]option.ClientOption, 0)
+
+	if o.Host != "" && o.Host != "/" {
+		fmt.Println("setting PUBSUB_EMULATOR_HOST as", o.Host)
+		os.Setenv("PUBSUB_EMULATOR_HOST", o.Host)
+	}
+
+	if o.ProjectID != "" {
+		fmt.Println("setting PUBSUB_PROJECT_ID as", o.ProjectID)
+		os.Setenv("PUBSUB_PROJECT_ID", o.ProjectID)
+	}
+
+	if o.JSONAuth != "" {
+		opts = append(opts, option.WithCredentialsFile(o.JSONAuth))
+	}
+
+	c = &Consumer{
+		info: info.Consumer{
+			Bus:   "pubsub",
+			Topic: o.Topic,
+		},
+	}
+
+	// create context for clean shutdown
+	c.ctx, c.cncl = context.WithCancel(context.Background())
+	c.client, err = pubsub.NewClient(c.ctx, o.ProjectID, opts...)
+	if err != nil {
+		return nil, err
+	}
+
+	// this loop will wait for the topic to be created (retry 10 times)
+	// if it cannot be created after the retries it will fail with a not exists
+	topic := c.client.Topic(o.Topic)
+	for retry := 0; ; retry++ {
+		ok, err := topic.Exists(c.ctx)
+		if err != nil {
+			return nil, err
+		}
+
+		if ok {
+			break
+		}
+
+		fmt.Printf("topic %s does not exist... waiting...", o.Topic)
```

Comment:
Lets remove the topic.Exists loop. This code doesn't actually create the topic, just waits for it to be created. If the topic doesn't exist, we should just get a failure

Reply:
So I updated this to create the topic if it doesn't exist. Should I change this to fail and exit if it doesn't exist?